src/System.Private.ServiceModel/tests/Scenarios/Extensibility/WebSockets/WebSocketTests.cs

```diff
@@ -598,4 +632,456 @@ public static void WebSocket_WSScheme_WSTransportUsageAlways_DuplexCallback_Guid
+        ClientReceiver clientReceiver = null;
+        InstanceContext context = null;
+        DuplexChannelFactory<IWSDuplexService> channelFactory = null;
+        IWSDuplexService client = null;
+
+        try
+        {
+            // *** SETUP *** \\
+            binaryMessageEncodingBindingElement = new BinaryMessageEncodingBindingElement();
+            httpsTransportBindingElement = new HttpsTransportBindingElement()
+            {
+                MaxReceivedMessageSize = ScenarioTestHelpers.DefaultMaxReceivedMessageSize,
+                MaxBufferSize = ScenarioTestHelpers.DefaultMaxReceivedMessageSize,
+            };
+            httpsTransportBindingElement.WebSocketSettings.TransportUsage = WebSocketTransportUsage.Always;
+            binding = new CustomBinding(binaryMessageEncodingBindingElement, httpsTransportBindingElement);
+
+            clientReceiver = new ClientReceiver();
+            context = new InstanceContext(clientReceiver);
+
+            channelFactory = new DuplexChannelFactory<IWSDuplexService>(context, binding, new EndpointAddress(Endpoints.WebSocketHttpsDuplexBinaryBuffered_Address));
+            client = channelFactory.CreateChannel();
+
+            // *** EXECUTE *** \\
+            // Invoking UploadData
+            client.UploadData(ScenarioTestHelpers.CreateInterestingString(123));
+
+            // Invoking StartPushingData
+            client.StartPushingData();
+            Assert.True(clientReceiver.ReceiveDataInvoked.WaitOne(ScenarioTestHelpers.TestTimeout),
+            "Test case timeout was reached while waiting for the buffered response from the Service.");
+            clientReceiver.ReceiveDataInvoked.Reset();
+            // Invoking StopPushingData
+            client.StopPushingData();
+            Assert.True(clientReceiver.ReceiveDataCompleted.WaitOne(ScenarioTestHelpers.TestTimeout),
+            "Test case timeout was reached while waiting for the buffered response from the Service to be completed.");
+            clientReceiver.ReceiveDataCompleted.Reset();
+
+            // Getting results from server via callback.
+            client.GetLog();
+            Assert.True(clientReceiver.LogReceived.WaitOne(ScenarioTestHelpers.TestTimeout),
+            "Test case timeout was reached while waiting for the Logging from the Service to be received.");
+
+            // *** VALIDATE *** \\
+            Assert.True(clientReceiver.ServerLog.Count > 0,
+                "The logging done by the Server was not returned via the Callback.");
+
+            // *** CLEANUP *** \\
+            ((ICommunicationObject)client).Close();
+            channelFactory.Close();
+
+        }
+        finally
+        {
+            // *** ENSURE CLEANUP *** \\  
+            ScenarioTestHelpers.CloseCommunicationObjects((ICommunicationObject)client, channelFactory);
+        }
+    }
+
+    [Fact]
+    [OuterLoop]
+    public static void WebSocket_Https_Duplex_TextBuffered_KeepAlive()
+    {
```

Comment:
I might be missing something... how are we testing that KeepAlive is working properly here?